### PR TITLE
Canny: cythonize non-maximum suppression

### PR DESCRIPTION
**Pull request by grlee77**
## Description

This PR refactors the non-maximum suppression stage of the Canny filter into Cython code. I think the new Cython function is more readable than `_get_local_maxima` and is also about 10x faster. Overall, the changes in this PR reduce the duration of the `time_canny` benchmark by approximately a factor of two.

Aside from the Cython code, some inefficiencies in `_preprocess` were addressed. Specifically, we don't need to compute the bleedover when there is no mask except in the case when mode='constant'.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Comment by lagru** [comment]
Thanks @grlee77 

> Specifically, we don't need to compute the bleedover when there is no mask except in the case when mode='constant'.

Can you explain the reason for this? Is it because the mask is !=1 outside the border only for the mode="constant" ? For all other modes, an empty mask defaults to 1 outside the border?

**Comment by alexdesiqueira** [comment]
Thanks @grlee77! I'm happy with this, but I'll ask @lagru's question, also :slightly_smiling_face: 
> Can you explain the reason for this? Is it because the mask is !=1 outside the border only for the mode="constant" ? For all other modes, an empty mask defaults to 1 outside the border?

**Comment by grlee77** [comment]
> > Specifically, we don't need to compute the bleedover when there is no mask except in the case when mode='constant'.

>? Can you explain the reason for this? Is it because the ma
sk is !=1 outside the border only for the mode="constant" ? For all other modes, an empty mask defaults to 1 outside the border?

Yes, when a user does not supply a mask, the following line just applies a Gaussian convolution to an mask that is 1 everywhere. So, regardless of whether we do mirroring, edge replication or periodic modes, all values outside the extent get set to 1 as well. It is only in the mode='constant' case where this is not true

```Python     
   bleed_over = gaussian(mask.astype(float_type, copy=False), **gaussian_kwargs)
```

**Comment by alexdesiqueira** [comment]
@lagru I'm happy with @grlee77's answer; let me know if you are as well, and we can merge this. Thanks!

File: skimage/feature/_canny.py [hunk 1]
```diff
@@ -319,31 +224,34 @@ def canny(image, sigma=1., low_threshold=None, high_threshold=None,
...
         return low_mask
 
-    high_mask = local_maxima & (magnitude >= high_threshold)
+    high_mask = low_mask & (low_masked >= high_threshold)
     nonzero_sums = np.unique(labels[high_mask])
     good_label = np.zeros((count + 1,), bool)
     good_label[nonzero_sums] = True
     output_mask = good_label[labels]
-    return output_mask
+    return output_mask
```
Comment (lagru) [inline skimage/feature/_canny.py:257]:
```suggestion
    return output_mask

```

File: skimage/feature/_canny_cy.pyx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                            neigh1_2 = magnitude[x - 1, y + 1]
+                            neigh2_1 = magnitude[x + 1, y]
+                            neigh2_2 = magnitude[x + 1, y - 1]
+                    # linear interpolation
+                    c_plus = (neigh1_2 * w + neigh1_1 * (1.0 - w)) <= m
+                    if c_plus:
+                        c_minus = (neigh2_2 * w + neigh2_1 * (1.0 - w)) <= m
+                        if c_minus:
+                            out[x, y] = m
+    return np.asarray(out)
```
Comment (lagru) [inline skimage/feature/_canny_cy.pyx:113]:
```suggestion
    return out.base
```

**Review by lagru — APPROVED** [review]
This looks good and I agree that the Cython version is significantly more readable. I wish there were a lot more comments in `_nonmaximum_suppression_bilinear`. I wasn't familiar with this algorithm at all so it took me some time digging through the old and new code as well as looking at resources [[ref1](https://link.springer.com/article/10.1007/s00180-020-01017-8#Sec2), [ref2](https://justin-liang.com/tutorials/canny/), ...]. Especially the interpolation could use a lot more explanation in comments and docstring. 

I'm approving this although I have some open suggestions because they are not that critical in the big picture and I don't want to stall this further. But I'd be happy about some clarification to "Question 1". That might help me add the explanations myself. ;)

Thanks @grlee77, the new Cython function is very cool!

**Comment by lagru** [comment]
Oh and sorry that I took so long to follow up. :pray:

**Comment by lagru** [comment]
Test failure seems unrelated to me.

**Comment by grlee77** [comment]
> I wish there were a lot more comments in _nonmaximum_suppression_bilinear

Yeah, I also found that code hard to follow. I think it is a little better now, but still a bit tricky. The following textbook has a diagram illustrating the linear interpolation in Fig. 8.12 that is similar to the blog post you linked above:

David Forsyth and Jean Ponce. Computer Vision: A Modern Approach, Prentice Hall, 2002.

My primary goal was to preserve the existing behavior for backwards compatibility. I also looked at the OpenCV implementation but it just uses a simpler nearest-neighbor approach instead of the linear interpolation approach used here.

File: skimage/feature/_canny.py [hunk 3]
```diff
@@ -319,31 +224,34 @@ def canny(image, sigma=1., low_threshold=None, high_threshold=None,
     # Gradient magnitude estimation
     jsobel = ndi.sobel(smoothed, axis=1)
     isobel = ndi.sobel(smoothed, axis=0)
-    magnitude = np.hypot(isobel, jsobel)
+    magnitude = isobel * isobel
+    magnitude += jsobel * jsobel
+    np.sqrt(magnitude, out=magnitude)
```
[the remark this answers is not in the document]
Comment (grlee77) [reply skimage/feature/_canny.py:229]: Yes, here is a quick example in ipython showing close to 3x improvement for size (1024, 1024) arrays.

```
In [2]: import numpy as np

In [3]: isobel = np.random.randn(1024, 1024).astype(np.float32)

In [4]: jsobel = np.random.randn(1024, 1024).astype(np.float32)

In [5]: %timeit np.hypot(isobel, jsobel)
4.61 ms ± 28.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: def _hypot(isobel, jsobel):
   ...:     magnitude = isobel * isobel
   ...:     magnitude += jsobel * jsobel
   ...:     np.sqrt(magnitude, out=magnitude)
   ...:     return magnitude
   ...: 

In [7]: %timeit _hypot(isobel, jsobel)
1.3 ms ± 7.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

File: skimage/feature/_canny_cy.pyx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                out[x, y] = 0
+                m = magnitude[x, y]
+                if not (eroded_mask[x, y] and (m >= low_threshold)):
+                    continue
+
+                is_down = (isobel[x, y] <= 0)
+                is_up = 1 - is_down
+
+                is_left = (jsobel[x, y] <= 0)
+                is_right = 1 - is_left
```
[the remark this answers is not in the document]
Comment (grlee77) [reply skimage/feature/_canny_cy.pyx:74]: 0 should be possible in regions of constant intensity. If both are 0 it is handled by the `m >= low_threshold`, but if only one or the other is then I think there might be possibility of a difference. It shouldn't hurt performance to change back to a >= comparison for `is_up` and `is_right` just to be sure.

**Comment by grlee77** [comment]
I made the final suggested changes and then merged. Thanks for reviewing this closely @lagru! The one CI failure was unrelated to the changes in this PR (it is a new deprecation warning in the SciPy 1.9 release candidate)

**Comment by lagru** [comment]
> David Forsyth and Jean Ponce. Computer Vision: A Modern Approach, Prentice Hall, 2002.

Thanks, I'll see if I can get access at my library here. If so I'll do a PR to add a little more documentation and this reference. :+1:
